### PR TITLE
[DISK_IO]Prefetch BlobAll for GetChunk request to reduce disk IOs

### PR DESCRIPTION
**Pull request by zzmao**
1. Prefetch BlobAll for GetChunk request, which only needs BlobData, to reduce disk IOs.
2. Refacotr CloudMessageReadSet to use ByteBuf as content store.

### CODE REVIEW

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java [hunk 1]
```diff
@@ -118,57 +119,49 @@ private void fetchDataFromReadSet() throws MessageFormatException {
...
-          logger.trace("Calculate offsets, read and verify storeKey time: {}",
-              SystemTime.getInstance().milliseconds() - startTime);
 
-          startTime = SystemTime.getInstance().milliseconds();
+          MessageHeader_Format headerFormat = parseHeaderAndVerifyStoreKey(bufferedInputStream, i);
+
+          long startTime = SystemTime.getInstance().milliseconds();
           if (flag == MessageFormatFlags.BlobProperties) {
             sendInfoList.add(i, new SendInfo(headerFormat.getBlobPropertiesRecordRelativeOffset(),
                 headerFormat.getBlobPropertiesRecordSize()));
```
Comment (cgtz) [inline ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java:167]: Can the `flag == Blob` case on line 203 be removed now?

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java [hunk 2]
```diff
@@ -2066,6 +2066,7 @@ static void endToEndReplicationWithMultiNodeSinglePartitionTest(String routerDat
       channel3.disconnect();
     } catch (Exception e) {
       e.printStackTrace();
+      System.out.println(e);
```
Comment (cgtz) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:2069]: extra println here

**Review by cgtz — COMMENTED** [review]
Nice, this doesn't seem to be too many changes to make this work!

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java (hunk 1 above)
Comment (zzmao) [reply ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java:167]: removed.

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java (hunk 2 above)
Comment (zzmao) [reply ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java]: fixed.

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java [hunk 3]
```diff
@@ -118,57 +119,49 @@ private void fetchDataFromReadSet() throws MessageFormatException {
...
-          totalSizeToWrite += readSet.sizeInBytes(i);
           readSet.doPrefetch(i, 0, readSet.sizeInBytes(i));
+
+          if (flag == MessageFormatFlags.All) {
+            sendInfoList.add(i, new SendInfo(0, readSet.sizeInBytes(i)));
+            messageMetadataList.add(i, null);
+            totalSizeToWrite += readSet.sizeInBytes(i);
+          } else if (flag == MessageFormatFlags.Blob) {
+            ByteBuf blobAll = readSet.getPrefetchedData(i);
+            InputStream is = new ByteBufInputStream(blobAll);
```
Comment (justinlin-linkedin) [inline ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java:134]: nit: we can use a more descriptive name than just `is`, something like `'blobInputStream`.

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java [hunk 4]
```diff
@@ -118,57 +119,49 @@ private void fetchDataFromReadSet() throws MessageFormatException {
...
+
+            MessageHeader_Format headerFormat = parseHeaderAndVerifyStoreKey(is, i);
+
+            MessageMetadata messageMetadata = null;
+            if (headerFormat.hasEncryptionKeyRecord()) {
+              // If encryption key exists, MessageMetadata with encryption key is needed.
+              ByteBuf duplicatedByteBuf = blobAll.duplicate();
+              duplicatedByteBuf.readerIndex(headerFormat.getBlobEncryptionKeyRecordRelativeOffset());
+              duplicatedByteBuf.writerIndex(headerFormat.getBlobEncryptionKeyRecordRelativeOffset()
```
Comment (justinlin-linkedin) [inline ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java:143]: you can do `duplicatedByteBuf.setIndex(readerIndex, writerIndex)` in just one statement.

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java [hunk 5]
```diff
@@ -118,57 +119,49 @@ private void fetchDataFromReadSet() throws MessageFormatException {
...
+                  new MessageMetadata(deserializeBlobEncryptionKey(new ByteBufInputStream(duplicatedByteBuf)));
+            }
+            messageMetadataList.add(messageMetadata);
+            sendInfoList.add(i,
+                new SendInfo(headerFormat.getBlobRecordRelativeOffset(), headerFormat.getBlobRecordSize()));
+            totalSizeToWrite += headerFormat.getBlobRecordSize();
+
+            // Adjust underlying ByteBuf reader and writer index.
+            blobAll.readerIndex(headerFormat.getBlobRecordRelativeOffset());
+            blobAll.writerIndex((int) (headerFormat.getBlobRecordRelativeOffset() + headerFormat.getBlobRecordSize()));
```
Comment (justinlin-linkedin) [inline ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java:155]: same above

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java [hunk 6]
```diff
@@ -118,57 +119,49 @@ private void fetchDataFromReadSet() throws MessageFormatException {
...
-          messageMetadataList.add(i, null);
-          totalSizeToWrite += readSet.sizeInBytes(i);
           readSet.doPrefetch(i, 0, readSet.sizeInBytes(i));
+
+          if (flag == MessageFormatFlags.All) {
+            sendInfoList.add(i, new SendInfo(0, readSet.sizeInBytes(i)));
+            messageMetadataList.add(i, null);
+            totalSizeToWrite += readSet.sizeInBytes(i);
+          } else if (flag == MessageFormatFlags.Blob) {
+            ByteBuf blobAll = readSet.getPrefetchedData(i);
```
Comment (justinlin-linkedin) [inline ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java:133]: I am a bit worried about the using prefetchedData here, without adding it to the `dataFromReadSet` list.

If there is any exception while parsing the header, this `ByteBuf` will be leaked. We should probably add the prefetchedData ByteBuf to the `dataFromReadSet` list before using it to parse the header.

**Review by justinlin-linkedin — COMMENTED** [review]
Didn't review the cloud MessageReadSet, just the MessageFormatSend. Most of it looks fine, just a couple nitpicking issues.

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java (hunk 3 above)
Comment (zzmao) [reply ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java]: sure. Changed.

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java (hunk 4 above)
Comment (zzmao) [reply ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java]: done.

File: ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java (hunk 6 above)
Comment (zzmao) [reply ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java:133]: `dataFromReadSet.add(readSet.getPrefetchedData(i));` also covers this case. (It's after the if-else for each message). no?

Comment (justinlin-linkedin) [reply ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java:133]: it does, but what if at line 136, parseHeaderAndVerifyStoreKey failed with an exception, then this ByteBuf will not be released.

Comment (zzmao) [reply ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java:134]: Got it. changes made.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java [hunk 7]
```diff
@@ -98,10 +97,12 @@ public StoreKey getKeyAt(int index) {
...
     try {
-      if (!blobReadInfo.isPrefetched()) {
-        blobReadInfo.prefetchBlob(blobStore);
+      if (!blobReadInfo.isBlobDownloaded()) {
+        blobReadInfo.downloadBlob(blobStore);
       }
-      blobReadInfo.setPositionAndSize(relativeOffset, size);
+      ByteBuf byteBuf = blobReadInfoList.get(index).getBlobContent();
+      byteBuf.readerIndex((int) (relativeOffset));
+      byteBuf.writerIndex((int) (relativeOffset + size));
```
Comment (justinlin-linkedin) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java:105]: nit: byteBuf.setIndex.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java [hunk 8]
```diff
@@ -140,81 +131,38 @@ private void validateIndex(int index) {
...
-    }
-
-    /**
-     * Donwload the blob from the {@code CloudDestination} to the {@code OutputStream}
-     * @param blobStore {@code CloudBlobStore} implementation representing the cloud from which download will happen.
-     * @param outputStream OutputStream to download the blob to
-     * @throws StoreException if blob download fails.
-     */
-    public void downloadBlob(CloudBlobStore blobStore, OutputStream outputStream) throws StoreException {
+      blobContent = PooledByteBufAllocator.DEFAULT.ioBuffer((int) blobMetadata.getSize());
```
Comment (justinlin-linkedin) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java:149]: is it safe to use netty ByteBuf in CloudMessageReadSet? @ankagrawal  what is the cloud network implementation? does it use http2? socket server?

**Review by justinlin-linkedin — COMMENTED** [review]
The MessageFormatSend.java file looks good though.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java (hunk 8 above)
Comment (zzmao) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java:149]: @justinlin-linkedin you are concerning if it's used in socket server, the ByteBuf can't be released?  
storeMessageSet had the same problem. I guess we should retire socket server soon.

Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java:149]: For vcr the network implementation currently is socket server based. It doesn't use http2.

Comment (ankagrawal) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java:149]: For now, until we have netty based http2 implementation in cloud, is there an issue with using `ByteBuffer.allocate` for cloud?

Comment (zzmao) [reply ambry-cloud/src/main/java/com/github/ambry/cloud/CloudMessageReadSet.java:149]: Both socket and http network stack can take care of allocated ByteBuf. As long as we have no cloud based network implementation, it's good. (Or release ByteBuf correctly in new implementations)

**Review by justinlin-linkedin — APPROVED** [review]

**Review by ankagrawal — APPROVED** [review]
lgtm!

**Pull request merged**